### PR TITLE
Rootless Kubeflow

### PR DESCRIPTION
**Pull request by juliusvonkohout**
**Which issue is resolved by this Pull Request:**
Resolves https://github.com/kubeflow/manifests/issues/2014

**Description of your changes:**

The main steps are adding an additional profile for istio-cni or ambient mesh, updating the documentation and manifest generation process. Then adding the baseline and restricted PSS as kustomize component to /contrib and extending the profile controller to annotate user namespaces with configurable PSS labels.

First Stage:

1. Implement Istio 1.17.5 and use it by default. This is important for the Kubeflow 1.8  feature freeze
2. Implement istio-cni (--set components.cni.enabled=true --set components.cni.namespace=kube-system) as second option.
3. Add simple tests similar to tests/gh-actions/install_istio.sh and tests/gh-actions/install_knative.sh for istio-cni and support both rootfull and rootless istio at the same time and give users one release to test

Second stage in a second PR:
4. Add pod security standards (https://kubernetes.io/docs/concepts/security/pod-security-standards/) base/restricted to manifests/contrib
5. Add istio-ambient to Kubeflow 1.9
6. Enforce PSS baseline (here you can still build OCI containers via Podman and buildah) in Kubeflow 1.9. It works with any istio
7. Optionally Enforce PSS restricted (this is where minor corner cases are affected) in Kubeflow 1.10

Third stage:
10. Upgrade Istio to 1.19 and use ambient service mesh by default in Kubeflow 1.10

**Checklist:**
- [ ] Unit tests pass:
  **Make sure you have installed kustomize == 3.2.1**
    1. `make generate-changed-only`
    2. `make test`

### CODE REVIEW

**Comment by juliusvonkohout** [comment]
> To clarify, this pull request is the rootful daemonset with rootless sidecar solution?

yes.

**Comment by juliusvonkohout** [comment]
@agilgur5 do you have any thoughts about this PR?

**Review by moorthy156 — COMMENTED** [review]
We are also running istio-cni to remove sidecar root requirement privileges, it is working fine us

**Comment by Talador12** [comment]
We ended up shifting to use a dedicated physical cluster instead of a virtual cluster for kubeflow, but we still want to try out this model to improve our security footprint. Looking forward to when this work is merged

**Comment by juliusvonkohout** [comment]
@moorthy156 do you want to help with an upstream implementation?

**Comment by kimwnasptd** [comment]
@juliusvonkohout what's the state of this PR? Is it ready for review?

My only concern with the CNI is the initContainer redirection, which could break KServe etc that also has init container that fetches data https://github.com/kubeflow/manifests/issues/2014#issuecomment-1036168983. 

What's the state with this?

**Comment by moorthy156** [comment]
> @moorthy156 do you want to help with an upstream implementation? If yes, please contact me on slack.

I would love too. Can i get directions on this

**Comment by MuhammadUsman90** [comment]
> > @moorthy156 do you want to help with an upstream implementation? If yes, please contact me on slack.
> 
> I would love too. Can i get directions on this

I would also like to be part of it. @juliusvonkohout @moorthy156  Can we arrange a small point on this?

**Comment by juliusvonkohout** [comment]
Yes, always reach out in the Kubeflow slack.

**Comment by juliusvonkohout** [comment]
TODO: Rebuild istio-cni-1-17 with only `-set components.cni.enabled=true --set components.cni.namespace=kube-system`, no additional flag such as `set values.pilot.autoscaleMin=1 --set values.gateways.istio-ingressgateway.autoscaleMin=1 --set components.cni.tag=1.17.5 --set values.global.proxy.resources.requests.cpu=10m --set tag=1.17.5`

**Comment by juliusvonkohout** [comment]
TODO: add our new tests to the github workflow similarl to the normal istio

**Comment by eslerm** [comment]
@juliusvonkohout could we please add SECURITY.md information similar to https://github.com/canonical/lxd/blob/main/SECURITY.md#what-qualifies-as-a-security-issue which describes Kubeflow's security context with and without the rootless feature? This can be tracked in a separate issue.

**Comment by juliusvonkohout** [comment]
@eslerm feel free to send me something adjusted to istio and I can commit it. Otherwise the upstream istio cni documentation mentioned here should be enough.

**Comment by tzstoyanov** [comment]
@kimwnasptd  can you take a look on that PR, should be ready for merge.  Thanks!

**Comment by annajung** [comment]
It would be easier to review if the PR is separated into
- 1.17.5 upgrade
- Istio-cni
- contrib components

Also for contrib, can you edit to follow https://github.com/kubeflow/manifests/blob/master/proposals/20220926-contrib-component-guidelines.md?

File: .github/workflows/kserve_kind_cni_test.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
+name: Build & Apply KServe manifests in KinD, using istio CNI
+on:
+  workflow_dispatch:
```
Comment (annajung) [inline .github/workflows/kserve_kind_cni_test.yaml:3]: There should be a trigger here based on changes to specific directories, not just workflow_dispatch.

File: common/istio-1-17/istio-install/base/patches/istiod-remove-pdb.yaml
Comment (annajung) [inline common/istio-1-17/istio-install/base/patches/istiod-remove-pdb.yaml:1]: this looks like it's just an empty file?

Comment (juliusvonkohout) [reply common/istio-1-17/istio-install/base/patches/istiod-remove-pdb.yaml:1]: not anymore :-)

File: .github/workflows/kserve_kind_cni_test.yaml (hunk 1 above)
Comment (juliusvonkohout) [reply .github/workflows/kserve_kind_cni_test.yaml:3]: @tzstoyanov will check this on monday

**Comment by juliusvonkohout** [comment]
> It would be easier to review if the PR is separated into
> 
>     * 1.17.5 upgrade
> 
>     * Istio-cni
> 
>     * contrib components
> 
> 
> Also for contrib, can you edit to follow https://github.com/kubeflow/manifests/blob/master/proposals/20220926-contrib-component-guidelines.md?

I will just remove /contrib from this PR for now.

**Comment by tzstoyanov** [comment]
@annajung comments are addressed, please can you take a look again

File: tests/gh-actions/install_knative-cni.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+set -euo pipefail
+echo "Installing KNative with isstio-cni ..."
```
Comment (annajung) [inline tests/gh-actions/install_knative-cni.sh:3]:
```suggestion
echo "Installing KNative with istio-cni ..."
```

File: common/istio-cni-1-17/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+In this section, we explain how to upgrade our istio kustomize packages
+by leveraging `istioctl`. Assuming the new version is `X.Y.Z` and the
+old version is `X1.Y1.Z1`:
+
+1.  Make a copy of the old istio manifests tree, which will become the
+    kustomization for the new Istio version:
+
+        $ export MANIFESTS_SRC=<path/to/manifests/repo>
+        $ export ISTIO_OLD=$MANIFESTS_SRC/common/istio-X1-Y1
+        $ export ISTIO_NEW=$MANIFESTS_SRC/common/istio-X-Y
```
Comment (annajung) [inline common/istio-cni-1-17/README.md:19]: README should be updated to point to istio-cni

File: .github/workflows/kserve_kind_cni_test.yaml (hunk 1 above)
Comment (annajung) [reply .github/workflows/kserve_kind_cni_test.yaml:7]: If I understand the trigger correctly, only people with write access will have access to trigger the workflow based on workflow_dispatch. So, not sure if it's necessary to keep

Comment (tzstoyanov) [reply .github/workflows/kserve_kind_cni_test.yaml:7]: Yes, but this would be the easiest way to run Istio CNI tests. Lets keep it at least for one release, until Istio CNI is not enabled by default. We can remove the manual trigger at any time.

Comment (annajung) [reply .github/workflows/kserve_kind_cni_test.yaml:7]: Should this test be triggered when there is a change to istio-cni-*? Can we trigger the test to ensure it runs successfully?

Comment (tzstoyanov) [reply .github/workflows/kserve_kind_cni_test.yaml]: Ok, that makes more sense than the current logic. Changed the trigger action, thanks Anna.

File: common/istio-cni-1-17/README.md (hunk 3 above)
Comment (juliusvonkohout) [reply common/istio-cni-1-17/README.md]: done

**Comment by annajung** [comment]
Thanks everyone! 
/lgtm

cc @kimwnasptd

**Comment by kimwnasptd** [comment]
Thanks for the review @annajung! I'll also try to take a look within the weekend and merge this

**Comment by mpaulgreen** [comment]
@juliusvonkohout  The knative serving and kubeflow pods are still stuck with isito-init failing as described in [Issue #2338](https://github.com/kubeflow/manifests/issues/2338). I was trying to run on Openshift 4.12.x and 4.13.x. 
`2023-09-04T11:38:02.320376Z	info	Istio iptables environment:
ENVOY_PORT=
INBOUND_CAPTURE_PORT=
ISTIO_INBOUND_INTERCEPTION_MODE=
ISTIO_INBOUND_TPROXY_ROUTE_TABLE=
ISTIO_INBOUND_PORTS=
ISTIO_OUTBOUND_PORTS=
ISTIO_LOCAL_EXCLUDE_PORTS=
ISTIO_EXCLUDE_INTERFACES=
ISTIO_SERVICE_CIDR=
ISTIO_SERVICE_EXCLUDE_CIDR=
ISTIO_META_DNS_CAPTURE=
INVALID_DROP=

2023-09-04T11:38:02.320411Z	info	Istio iptables variables:
PROXY_PORT=15001
PROXY_INBOUND_CAPTURE_PORT=15006
PROXY_TUNNEL_PORT=15008
PROXY_UID=1337
PROXY_GID=1337
INBOUND_INTERCEPTION_MODE=REDIRECT
INBOUND_TPROXY_MARK=1337
INBOUND_TPROXY_ROUTE_TABLE=133
INBOUND_PORTS_INCLUDE=*
INBOUND_PORTS_EXCLUDE=15090,15021,15020
OUTBOUND_OWNER_GROUPS_INCLUDE=*
OUTBOUND_OWNER_GROUPS_EXCLUDE=
OUTBOUND_IP_RANGES_INCLUDE=*
OUTBOUND_IP_RANGES_EXCLUDE=
OUTBOUND_PORTS_INCLUDE=
OUTBOUND_PORTS_EXCLUDE=
KUBE_VIRT_INTERFACES=
ENABLE_INBOUND_IPV6=false
DNS_CAPTURE=false
DROP_INVALID=false
CAPTURE_ALL_DNS=false
DNS_SERVERS=[],[]
OUTPUT_PATH=
NETWORK_NAMESPACE=
CNI_MODE=false
HOST_NSENTER_EXEC=false
EXCLUDE_INTERFACES=

2023-09-04T11:38:02.320573Z	info	Writing following contents to rules file: /tmp/iptables-rules-1693827482320445198.txt2641253036
* nat
-N ISTIO_INBOUND
-N ISTIO_REDIRECT
-N ISTIO_IN_REDIRECT
-N ISTIO_OUTPUT
-A ISTIO_INBOUND -p tcp --dport 15008 -j RETURN
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A ISTIO_INBOUND -p tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15021 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
COMMIT
2023-09-04T11:38:02.320602Z	info	Running command: iptables-restore --noflush /tmp/iptables-rules-1693827482320445198.txt2641253036
2023-09-04T11:38:02.321779Z	error	Command error output: xtables parameter problem: iptables-restore: unable to initialize table 'nat'

Error occurred at line: 1
Try `iptables-restore -h' or 'iptables-restore --help' for more information.
2023-09-04T11:38:02.321791Z	error	Failed to execute: iptables-restore --noflush /tmp/iptables-rules-1693827482320445198.txt2641253036, exit status 2`

**Comment by mpaulgreen** [comment]
@juliusvonkohout You are correct. There are some additional configuration required istio-cni. I will stop posting defects of kubeflow with istio here.

**Comment by kimwnasptd** [comment]
Tested as well in a KinD cluster and everything worked as expected. Great to see this in, thank you everyone that worked hard!

/lgtm
/approve
